Supporting Date spec in File Based ExampleGen driver

18jeffreyma opened the pull request:
Currently, in FileBasedExampleGen, users must specify span explicitly with the `{SPAN}` tag; however, users may have data organized by date stamp, with This CL adds support for parsing the calendar date (year, month, and day) of data with a set of Date specs `{YYYY}, {MM}, {DD}`. Using this calendar date, the span number is calculated as the number of days since unix epoch (`1970-01-01`).

1025KB commented on tfx/components/example_gen/utils.py:
```diff
@@ -346,12 +393,18 @@ def calculate_splits_fingerprint_span_and_version(
                  select_version)
     # Find most recent span and version for this split.
-    latest_span, latest_version = _retrieve_latest_span_version(
+    latest_span, latest_span_elems, latest_version, = _retrieve_latest_span_elems_version(
```
my bad, i feel we should probably move split.pattern change in _retrieve_latest_span_elems_version, and just return span and version (to align, both span and version should be int, we might need to support {span:2} for the format (`01`, `02`), can be in future PR)

18jeffreyma replied:
should we just be setting them as int artifact properties?

18jeffreyma replied:
updated PR to move split.pattern update inside `_retrieve_latest_span_version()`

18jeffreyma replied:
also updated return type to be ints, and driver to set span and version as int properties

1025KB commented on tfx/components/example_gen/utils.py:
```diff
@@ -300,19 +332,34 @@ def _retrieve_latest_span_version(
 
-    if latest_span is None or span_int > int(latest_span):
+    if (latest_span_elems is None
+        or span_elems_int > [int(elem) for elem in latest_span_elems]):
```
compare two list seems not ideal,  convert the date to span_int and still use span_int for comparison?

18jeffreyma replied:
fixed. now does conversion for each file path

1025KB commented on tfx/components/example_gen/utils.py:
```diff
@@ -283,9 +315,9 @@ def _retrieve_latest_span_version(
       raise ValueError('Glob pattern does not match regex pattern')
 
-    span_str = result.group(SPAN_PROPERTY_NAME)
+    span_elems_str = [result.group(name) for name in span_group_names]
```
I feel have the span and date branch here might be more readable instead of using the magic `span_group_names ` and line348~360 

```
if is_match_span:
  span_strs = [...]
  span_int =
elif is_match_date:
  span_strs = 
  span_int = 

if span_int > xxx
```

18jeffreyma replied:
refactored to have more clean branching logic.

18jeffreyma commented:
Resolved latest round of feedback thanks! Noteable changes:
- Made logic more clear by separating `is_match_span` and `is_match_date` from glob loop.
- Split pattern now updated from inside `_retrieve_latest_span_elems_version()`

casassg commented:
Could we add support for instead of only resolving with latest available date span, to resolve to the latest span available before Airflow execution_date (when using Airflow as scheduler). This would unlock using Airflow native backfill to generate datasets as they would have been available in X date.

Use case:
- I want to fo data validation on the dataset I trained my current outdated production model

18jeffreyma commented:
@casassg This PR is primarily for adding Date spec (and logic to parse Date spec) to pull latest data in FileBasedExampleGen. I believe support for processing specific (non-most recent) spans (and/or dates), will be added in a later PR.

casassg commented:
Sounds good. Looking forward to that one then. Will try to keep an eye

18jeffreyma commented:
More updates:
- Span and Version are now returned as integers, and they are set as custom integer properties for Example artifacts.

1025KB commented on RELEASE.md:
```diff
@@ -44,6 +44,7 @@
 *   Added custom config to Transform component and made it available to
     pre-processing fn.
 *   Supported custom extractors in Evaluator.
+*   Supported `Date` spec in input config for file based ExampleGen.
```
mention the type is changed in breaking change

18jeffreyma replied:
updated, thanks!

ruoyu90 replied:
I didn't find anything added under '## Breaking changes' section?

18jeffreyma replied:
@ruoyu90 we decided to revert this type change. the proposed change was making span and version int custom properties, but since that would likely block the CL, we've decided to push it back.

https://github.com/tensorflow/tfx/pull/2285#discussion_r469564944

18jeffreyma commented:
Resolved latest round of feedback, thanks!

1025KB commented on tfx/components/example_gen/utils_test.py:
```diff
@@ -427,8 +518,66 @@ def testCalculateSplitsFingerprintSpanAndVersion(self):
+    _, span, version = utils.calculate_splits_fingerprint_span_and_version(
+        self._input_base_path, splits)
+    self.assertEqual(span, 2)
+    self.assertEqual(version, 2)
```
is there a test that verify the split pattern updates of calculate_splits_fingerprint_span_and_version?

we want to make sure it's covered (especially for format like '02')

18jeffreyma replied:
added tests to check for updated split patterns

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -225,13 +234,19 @@ def _retrieve_latest_span_version(
 
   Returns:
-    Tuple of two strings, Span (optional) and Version (optional).
+    Tuple of two strings, Span (optional) and Version (optional). Note
```
nit, tuple of two int

18jeffreyma replied:
fixed, thanks!

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -241,27 +256,50 @@ def _retrieve_latest_span_version(
+                     'exclusively.')
 
   split_pattern = os.path.join(uri, split.pattern)
 
   split_glob_pattern = split_pattern
   split_regex_pattern = _glob_to_regex(split_pattern)
 
-  latest_span = None
-  latest_version = None
-
-  if SPAN_SPEC not in split.pattern:
+  if not is_match_span and not is_match_date:
     if VERSION_SPEC in split.pattern:
-      raise ValueError('Version spec provided, but Span spec is not present')
-    return latest_span, latest_version
+      raise ValueError('Version spec provided, but Span or Date spec is not '
+                       'present.')
+    return (None, None)
 
-  if split.pattern.count(SPAN_SPEC) != 1:
+  if is_match_span and split.pattern.count(SPAN_SPEC) != 1:
     raise ValueError('Only one %s is allowed in %s' %
                      (SPAN_SPEC, split.pattern))
+  elif is_match_date and not all(split.pattern.count(spec) == 1
+                                 for spec in DATE_SPECS):
+    raise ValueError('Exactly one of each date spec is required in %s' %
+                     split.pattern)
```
nit, consider move all the validation to to some util

18jeffreyma replied:
fixed. moved all initial spec finding validation into `_verify_split_specs()`, which returns bools indicating what should be matched

hughmiao commented on tfx/components/example_gen/driver.py:
```diff
@@ -92,11 +92,11 @@ def _prepare_output_artifacts(
         utils.FINGERPRINT_PROPERTY_NAME,
         exec_properties[utils.FINGERPRINT_PROPERTY_NAME])
-    example_artifact.set_string_custom_property(
+    example_artifact.set_int_custom_property(
```
note, this is breaking change, we shall still use string custom property for this?

/cc jyzhao@ any context of changing this to int?

18jeffreyma replied:
@1025KB

1025KB replied:
Hui, we change to int because internal we use int?

or you suggest to keep the string custom property and later we add int property(not custom property) directly?

hughmiao replied:
>  you suggest to keep the string custom property and later we add int property(not custom property) directly?

that looks a better approach. I think to unblock the cl, let's keep using string for now, and we can always do it in separate cls. wdyt?

18jeffreyma replied:
sgtm. reverted back to using `set_string_custom_property()`. util functions like `calculate_splits_fingerprint_span_and_version()` still return ints, but `_prepare_output_artifacts()` in `driver.py` coverts int back to string when setting artifact property.

18jeffreyma commented:
resolved latest round of feedback, thanks! notable changes:
- split pattern spec preprocessing (and error raising) is now handled in a private function.
- span and version have been reverted back to string custom properties for example artifact

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -214,9 +223,39 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+  if is_match_span and split.pattern.count(SPAN_SPEC) != 1:
+    raise ValueError('Only one %s is allowed in %s' % (SPAN_SPEC,
+                                                       split.pattern))
+  elif is_match_date and not all(split.pattern.count(spec) == 1
```
instead of `elif`, use a separate `if` ?

18jeffreyma replied:
fixed

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -214,9 +223,39 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+                                                       split.pattern))
+  elif is_match_date and not all(split.pattern.count(spec) == 1
+                                 for spec in DATE_SPECS):
+    raise ValueError('Exactly one of each date spec is required in %s' %
```
nit, add YYYY, MM, DD to the error string, which helps the users to fix their config

18jeffreyma replied:
updated

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -214,9 +223,39 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+
+  if [is_match_span, is_match_date].count(True) > 1:
+    raise ValueError('Either span spec or date specs must be specified '
+                     'exclusively.')
```
add the split.pattern to the error string like other cases, which helps users to fix configs

18jeffreyma replied:
fixed

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -214,9 +223,39 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+
+  if is_match_version and (not is_match_span and not is_match_date):
+    raise ValueError('Version spec provided, but Span or Date spec is not '
+                      'present.')
```
ditto, + split.pattern

18jeffreyma replied:
updated.

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -276,19 +322,40 @@ def _retrieve_latest_span_version(
+                         (SPAN_PROPERTY_NAME, file_path, split_regex_pattern))
+
+    elif is_match_date:
+      span_strs = [result.group(name) for name in ['year', 'month', 'day']]
```
nit, for ['year', 'month', 'day'], using some constants?

18jeffreyma replied:
constraints are handled by the `datetime` package

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -276,19 +322,40 @@ def _retrieve_latest_span_version(
-    except ValueError:
-      raise ValueError('Cannot find %s number from %s based on %s' %
-                       (SPAN_PROPERTY_NAME, file_path, split_regex_pattern))
+    span_strs = None
```
nit, s/span_strs/matched_span_tokens

18jeffreyma replied:
updated. also changed `latest_span_elems` to `latest_span_tokens`

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -300,46 +367,60 @@ def _retrieve_latest_span_version(
+    split.pattern = split.pattern.replace(VERSION_SPEC, latest_version)
+    latest_version_int = int(latest_version)
+
+  return latest_span_int, latest_version_int
```
nit, the function still a bit long, probably decompose it to logical steps improve readability, 

e.g., 
_find_matched_span_tokens_from_path
_resolve_latestest_span_version

18jeffreyma replied:
addressed. decomposed span matching as well as glob and regex generation to two different private functions.

hughmiao commented on tfx/components/example_gen/utils.py:
```diff
@@ -300,46 +367,60 @@ def _retrieve_latest_span_version(
-    splits: An iterable collection of example_gen_pb2.Input.Split objects. Note
-      that this function will update the {SPAN} in this and {VERSION} tags in
-      the split config to actual Span and Version numbers.
+    splits: An iterable collection of example_gen_pb2.Input.Split objects.
```
add the Note the tags will be updated part in the previous comment

18jeffreyma replied:
updated, thanks!

18jeffreyma commented on RELEASE.md:
```diff
@@ -45,6 +45,10 @@
 *   Supported custom extractors in Evaluator.
 *   Deprecated tensorflow dependency from MLMD python client.
+*   Supported `Date` spec in input config for file based ExampleGen.
+*   Span and Version are now stored as `int` custom properties for example
+    artifacts, where previously they were stored as strings.
```
apologies forgot to remove on revert, fixed.

18jeffreyma commented:
addressed above comments, thanks!

18jeffreyma commented on tfx/components/example_gen/utils.py:
```diff
@@ -214,9 +223,123 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+def _find_matched_span_version_from_path(
+    file_path: Text, split_regex_pattern: Text, is_match_span: bool,
+    is_match_date: bool, is_match_version: bool
+) -> Tuple[Optional[List[Text]], Optional[int]]:
```
update typing
